src/d/parser/expression.d

```diff
@@ -1047,3 +1050,85 @@ StringLiteral parseStringLiteral(ref TokenRange trange) {
 	
 	return new StringLiteral(location, name.toString(trange.context));
 }
+
+/**
+ * Parse floating point literals
+ */
+FloatLiteral parseFloatLiteral(ref TokenRange trange) {
+	const location = trange.front.location;
+	auto litString = trange.front.toString(trange.context);
+
+	trange.match(TokenType.FloatLiteral);
+	import d.common.builtintype : BuiltinType;
+
+	//Look for a suffix
+	if (litString.length > 1)
+	{
+		switch (litString[$-1])
+		{
+			// https://dlang.org/spec/lex.html#FloatSuffix
+			import std.conv : to;
+			case 'f':
+			case 'F':
+				const float f = litString[0..$-1].to!float;
+				return new FloatLiteral(location, f, BuiltinType.Float);
+			case 'L':
+				import source.exception;
+				throw new CompileException(location, "SDC does not support real literals yet");
+			default:
+				//Lexed correctly but no suffix, it's a double
+				const double d = litString[0..$].to!double;
+				return new FloatLiteral(location, d, BuiltinType.Double);
+		}
+	}
+	assert(0);
+}
+
+@("Test FloatLiteral parsing")
+unittest
+{
+	import source.context;
+	auto context = new Context();
+	import source.parserutil;
+	auto tokensFromString(string s) {
+		import source.name;
+		auto base = context.registerMixin(Location.init, s ~ '\0');
+		auto x = lex(base, context);
+		x.match(TokenType.Begin);
+		return x;
+	}
+	void floatRoundTrip(T)(const string floatString, const T floatValue)
+		if (__traits(isFloating, T))
+	{
+		import d.common.builtintype : BuiltinType;
+		const BuiltinType expectedType = is(T == float) ? BuiltinType.Float : BuiltinType.Double;
+		//Acceptable relativeError
+		const auto maxRelError = is(T == float) ? float.epsilon : double.epsilon;
+		auto tr = tokensFromString(floatString);
+		const result = parseFloatLiteral(tr);
+		import std.format : format;
+		if (const FloatLiteral fl = cast(FloatLiteral) result)
```

Comment:
```d
const FloatLiteral fl = cast(FloatLiteral) result;
assert(fl, format("Got a %s from `%s`", typeid(result).toString(), floatString));

// Do the rest here
```

Don't overnest code when it is not needed. When reading nested code, one has to keep a lot of context in their mind about all the contexts that are entered, but it often isn't needed.

Reply:
Done